Mathlib/Tactic/Ring/Basic.lean

```diff
@@ -987,6 +987,26 @@ theorem div_congr {R} [DivisionRing R] {a a' b b' c : R} (_ : a = a') (_ : b = b
 /-- A precomputed `Cache` for `ℕ`. -/
 def Cache.nat : Cache sℕ := { rα := none, dα := none, czα := some q(inferInstance) }
 
+/-- Checks whether `e` will be treated as an atom by `eval`. -/
+-- Note this is not the same as whether the result of `eval` is an atom. (e.g. consider `x + 0`.)
+def isAtom {u} {α : Q(Type u)} (sα : Q(CommSemiring $α))
+    (c : Cache sα) (e : Q($α)) : AtomM Bool := do
+  let els : MetaM Bool := try
+      _ ← derive e
```

Comment:
Hm, this is unfortunate. How about a lighter version of `NormNum.derive` which just checks whether there are any applicable rules in the NormNum discrimination tree? Alternatively, we could return the result of `derive` from this function and apply it to the expression. The worry is that we have to evaluate any numerics twice this way.

Reply:
I've made another attempt. This time, the lightweight function is `isAtomOrDerivable`, which either returns:
* `none` if the expression "looks algebraic enough for `eval`"
* `some (some r)` otherwise, if `NormNum.derive` can simplify the expression
* `some none` otherwise, in which case we can consider the expression an atom.

Then, in `RingNF.rewrite`, we call this first. In the `some none` (i.e. atom) case, we just stop processing. In the `some (some r)` case, we just use that result immediately, without calling `eval` again. In the `none` case, we actually call `eval`.

I think this avoids evaluating any numerics twice, because if we are in the `none` case, then we know that `eval` is going to handle the outermost algebraic expression itself (i.e. by constructing a ring normal form), rather than passing it to `NormNum.derive`.